packages/fastify-podlet-server/lib/fastify-podlet-plugin.js

```diff
@@ -50,556 +50,575 @@ const resolveAssetsBasePath = ({ base, prefix, fallback }) => {
+  }
+
+  /**
+   * Sets up response timing metrics which hooks into Fastify via a plugin and exposes
+   * metrics via a stream which we then gather up into this.#metricStreams
+   */
+  timingMetrics() {
+    const responseTiming = new ResponseTiming({
+      timeAllRoutes: this.#timeAllRoutes,
+      groupStatusCodes: this.#groupStatusCodes,
     });
+    this.#fastify.register(responseTiming.plugin());
+    this.#metricStreams.push(responseTiming.metrics);
   }
 
   /**
-   * Collect up all metrics and expose on a .metrics property of the fastify instance
-   * which can then be piped into a consumer in server.js
+   * Sets up metric stream processing. Must be called at the end after other modules have
+   * had a chance to push metric stream objects into this.#metricStreams after which
+   * we gather up these streams and pipe them altogether and expose a single metric object
+   * to be used externally for consumption
    */
-  if (METRICS_ENABLED) {
-    const metrics = new Metrics();
-    for (const stream of metricStreams) {
+  metricStreams() {
+    /**
+     * Collect up all metrics and expose on a .metrics property of the fastify instance
+     * which can then be piped into a consumer in server.js
+     */
+    for (const stream of this.#metricStreams) {
       stream.on("error", (err) => {
-        fastify.log.error(err);
+        this.#logger.error(err);
       });
-      stream.pipe(metrics);
+      stream.pipe(this.#metrics);
     }
-    fastify.decorate("metrics", metrics);
   }
-};
+
+  /**
+   * Sets up and returns a fastify plugin with everything necessary for a running app.
+   * @returns {import("fastify").FastifyPluginAsync}
+   */
+  plugin() {
+    // return plugin
+    return async (fastify) => {
+      this.#fastify = fastify;
+
+      this.podletPlugin();
+      await this.manifestRoute();
+      this.translations();
+      await this.compression();
+      await this.processExceptionHandlers();
+      this.timingMetrics();
+
+      if (this.#config.get("app.component")) {
+        await this.contentRoute();
+        await this.fallbackRoute();
+        await this.dependenciesRoute();
+        await this.serveAssets();
+      }
+
+      this.metricStreams();
+    };
+  }
+}
 
 export default fp(async function (fastify, { config }) {
-  fastify.register(plugin, { prefix: config.get("app.base") || "/", config });
+  const prefix = config.get("app.base") || "/";
+  const podletServer = new PodletServerPlugin({ config, logger: fastify.log, prefix });
+
+  fastify.register(podletServer.plugin(), { prefix });
+
+  // Expose developer facing APIs using decorate
+  fastify.decorate("setContentState", podletServer.setContentState.bind(podletServer));
+  fastify.decorate("setFallbackState", podletServer.setFallbackState.bind(podletServer));
+  fastify.decorate("config", podletServer.config);
+  fastify.decorate("podlet", podletServer.podlet);
+  fastify.decorate("metrics", podletServer.metrics);
```

Comment:
using the class based approach, we are able to decorate the outer plugin with values that are set from within the inner plugin.